tcmur_cmd_handler.c

```diff
@@ -456,6 +456,134 @@ static int handle_read(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
 	return async_handle_cmd(dev, cmd, read_work_fn);
 }
 
+/* format unit */
+struct format_unit_state
+{
+	size_t length;
+	off_t offset;
+	void *write_buf;
+	struct tcmulib_cmd *origcmd;
+	uint32_t done_blocks;
+};
+
+static int format_unit_work_fn(struct tcmu_device *dev,
+			       struct tcmulib_cmd *writecmd)
+{
+	struct tcmulib_handler *handler = tcmu_get_dev_handler(dev);
+	struct tcmur_handler *rhandler = handler->hm_private;
+	struct tcmulib_cmd *origcmd = writecmd->cmdstate;
+	struct format_unit_state *state = origcmd->cmdstate;
+
+	return rhandler->write(dev, writecmd, writecmd->iovec,
+			       writecmd->iov_cnt, state->length, state->offset);
+}
+
+static void handle_format_unit_cbk(struct tcmu_device *dev,
+				   struct tcmulib_cmd *writecmd, int ret)
+{
+	struct tcmulib_cmd *origcmd = writecmd->cmdstate;
+	struct format_unit_state *state = origcmd->cmdstate;
+	uint8_t *sense = origcmd->sense_buf;
+
+	writecmd->iovec->iov_base = state->write_buf;
+	state->offset += state->length;
+	state->done_blocks += state->length / dev->block_size;
+	if (state->done_blocks < dev->num_lbas) {
+		if ((dev->num_lbas - state->done_blocks) * dev->block_size < state->length) {
+			state->length = (dev->num_lbas - state->done_blocks) * dev->block_size;
+			free_iovec(writecmd);
+			if (alloc_iovec(writecmd, state->length)) {
+				ret = tcmu_set_sense_data(sense, HARDWARE_ERROR,
+							  ASC_INTERNAL_TARGET_FAILURE,
+							  NULL);
+				goto free_cmd;
```

Comment:
I go to free_cmd which clears the formatting flag. Also i will probably just leave it at allocation/free since I dont know consequences of hacking iovec's iov_len. That was my first consideration though.